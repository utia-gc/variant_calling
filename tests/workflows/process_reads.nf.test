nextflow_workflow {

    name "Test Workflow PROCESS_READS"
    script "workflows/process_reads.nf"
    workflow "PROCESS_READS"

    test("PROCESS_READS trims reads by default -- cutadapt + SE reads.") {
        // instantiate helper objects
        def readsLane1 = new ReadsSELane1()
        def readsLane2 = new ReadsSELane2()

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
                tools {
                    trim = 'cutadapt'
                }
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [
                            sampleName:   "${readsLane1.getSampleName()}",
                            sampleNumber: "${readsLane1.getSampleNumber()}",
                            lane:         "${readsLane1.getLane()}",
                            readType:     "${readsLane1.getReadType()}"
                        ],
                        file("${readsLane1.getR1()}"),
                        file("${readsLane1.getR2()}")
                    ],
                    [
                        [
                            sampleName:   "${readsLane2.getSampleName()}",
                            sampleNumber: "${readsLane2.getSampleNumber()}",
                            lane:         "${readsLane2.getLane()}",
                            readType:     "${readsLane2.getReadType()}"
                        ],
                        file("${readsLane2.getR1()}"),
                        file("${readsLane2.getR2()}")
                    ]
                )
                """
            }
        }

        then {
            // test workflow status
            assert workflow.success
            assert workflow.trace.succeeded().size() == 2

            // test read trimming and concatenation
            with(workflow.out.reads_pre_align) {
                assert size() == 2
                with(get(0)) {
                    assert size() == 3
                    assert get(1) ==~ /^.*\/${readsLane1.getStemName()}_trimmed_R1\.fastq\.gz/
                    assert get(2) ==~ /^.*\/${readsLane1.getStemName()}_trimmed_R2\.NOFILE/
                }
                with(get(1)) {
                    assert size() == 3
                    assert get(1) ==~ /^.*\/${readsLane2.getStemName()}_trimmed_R1\.fastq\.gz/
                    assert get(2) ==~ /^.*\/${readsLane2.getStemName()}_trimmed_R2\.NOFILE/
                }
            }
            assert snapshot(workflow.out.reads_pre_align).match()

            // test read trimming log
            with(workflow.out.trim_log) {
                assert size() == 2
                assert get(0) ==~ /^.*\/${readsLane1.getStemName()}_cutadapt-log\.txt/
                assert get(1) ==~ /^.*\/${readsLane2.getStemName()}_cutadapt-log\.txt/
            }
        }

    }


    test("PROCESS_READS trims reads by default -- cutadapt + PE reads.") {
        // instantiate helper objects
        def readsLane1 = new ReadsPELane1()
        def readsLane2 = new ReadsPELane2()

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
                tools {
                    trim = 'cutadapt'
                }
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [
                            sampleName:   "${readsLane1.getSampleName()}",
                            sampleNumber: "${readsLane1.getSampleNumber()}",
                            lane:         "${readsLane1.getLane()}",
                            readType:     "${readsLane1.getReadType()}"
                        ],
                        file("${readsLane1.getR1()}"),
                        file("${readsLane1.getR2()}")
                    ],
                    [
                        [
                            sampleName:   "${readsLane2.getSampleName()}",
                            sampleNumber: "${readsLane2.getSampleNumber()}",
                            lane:         "${readsLane2.getLane()}",
                            readType:     "${readsLane2.getReadType()}"
                        ],
                        file("${readsLane2.getR1()}"),
                        file("${readsLane2.getR2()}")
                    ]
                )
                """
            }
        }

        then {
            // test workflow status
            assert workflow.success
            assert workflow.trace.succeeded().size() == 2

            // test read trimming and concatenation
            with(workflow.out.reads_pre_align) {
                assert size() == 2
                with(get(0)) {
                    assert size() == 3
                    assert get(1) ==~ /^.*\/${readsLane1.getStemName()}_trimmed_R1\.fastq\.gz/
                    assert get(2) ==~ /^.*\/${readsLane1.getStemName()}_trimmed_R2\.fastq\.gz/
                }
                with(get(1)) {
                    assert size() == 3
                    assert get(1) ==~ /^.*\/${readsLane2.getStemName()}_trimmed_R1\.fastq\.gz/
                    assert get(2) ==~ /^.*\/${readsLane2.getStemName()}_trimmed_R2\.fastq\.gz/
                }
            }
            assert snapshot(workflow.out.reads_pre_align).match()

            // test read trimming log
            with(workflow.out.trim_log) {
                assert size() == 2
                assert get(0) ==~ /^.*\/${readsLane1.getStemName()}_cutadapt-log.txt/
                assert get(1) ==~ /^.*\/${readsLane2.getStemName()}_cutadapt-log.txt/
            }
        }

    }


    test("PROCESS_READS trims reads by default -- fastp + SE reads.") {
        // instantiate helper objects
        def readsLane1 = new ReadsSELane1()
        def readsLane2 = new ReadsSELane2()

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
                tools {
                    trim = 'fastp'
                }
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [
                            sampleName:   "${readsLane1.getSampleName()}",
                            sampleNumber: "${readsLane1.getSampleNumber()}",
                            lane:         "${readsLane1.getLane()}",
                            readType:     "${readsLane1.getReadType()}"
                        ],
                        file("${readsLane1.getR1()}"),
                        file("${readsLane1.getR2()}")
                    ],
                    [
                        [
                            sampleName:   "${readsLane2.getSampleName()}",
                            sampleNumber: "${readsLane2.getSampleNumber()}",
                            lane:         "${readsLane2.getLane()}",
                            readType:     "${readsLane2.getReadType()}"
                        ],
                        file("${readsLane2.getR1()}"),
                        file("${readsLane2.getR2()}")
                    ]
                )
                """
            }
        }

        then {
            // test workflow status
            assert workflow.success
            assert workflow.trace.succeeded().size() == 2

            // test read trimming and concatenation
            with(workflow.out.reads_pre_align) {
                assert size() == 2
                with(get(0)) {
                    assert size() == 3
                    assert get(1) ==~ /^.*\/${readsLane1.getStemName()}_trimmed_R1\.fastq\.gz/
                    assert get(2) ==~ /^.*\/${readsLane1.getStemName()}_trimmed_R2\.NOFILE/
                }
                with(get(1)) {
                    assert size() == 3
                    assert get(1) ==~ /^.*\/${readsLane2.getStemName()}_trimmed_R1\.fastq\.gz/
                    assert get(2) ==~ /^.*\/${readsLane2.getStemName()}_trimmed_R2\.NOFILE/
                }
            }
            assert snapshot(workflow.out.reads_pre_align).match()

            // test read trimming log
            with(workflow.out.trim_log) {
                assert size() == 2
                assert get(0) ==~ /^.*\/${readsLane1.getStemName()}_fastp.json/
                assert get(1) ==~ /^.*\/${readsLane2.getStemName()}_fastp.json/
            }
        }

    }


    test("PROCESS_READS trims reads by default -- fastp +  PE reads.") {
        // instantiate helper objects
        def readsLane1 = new ReadsPELane1()
        def readsLane2 = new ReadsPELane2()

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
                tools {
                    trim = 'fastp'
                }
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [
                            sampleName:   "${readsLane1.getSampleName()}",
                            sampleNumber: "${readsLane1.getSampleNumber()}",
                            lane:         "${readsLane1.getLane()}",
                            readType:     "${readsLane1.getReadType()}"
                        ],
                        file("${readsLane1.getR1()}"),
                        file("${readsLane1.getR2()}")
                    ],
                    [
                        [
                            sampleName:   "${readsLane2.getSampleName()}",
                            sampleNumber: "${readsLane2.getSampleNumber()}",
                            lane:         "${readsLane2.getLane()}",
                            readType:     "${readsLane2.getReadType()}"
                        ],
                        file("${readsLane2.getR1()}"),
                        file("${readsLane2.getR2()}")
                    ]
                )
                """
            }
        }

        then {
            // test workflow status
            assert workflow.success
            assert workflow.trace.succeeded().size() == 2

            // test read trimming and concatenation
            with(workflow.out.reads_pre_align) {
                assert size() == 2
                with(get(0)) {
                    assert size() == 3
                    assert get(1) ==~ /^.*\/${readsLane1.getStemName()}_trimmed_R1.fastq.gz/
                    assert get(2) ==~ /^.*\/${readsLane1.getStemName()}_trimmed_R2.fastq.gz/
                }
                with(get(1)) {
                    assert size() == 3
                    assert get(1) ==~ /^.*\/${readsLane2.getStemName()}_trimmed_R1.fastq.gz/
                    assert get(2) ==~ /^.*\/${readsLane2.getStemName()}_trimmed_R2.fastq.gz/
                }
            }
            assert snapshot(workflow.out.reads_pre_align).match()

            // test read trimming log
            with(workflow.out.trim_log) {
                assert size() == 2
                assert get(0) ==~ /^.*\/${readsLane1.getStemName()}_fastp.json/
                assert get(1) ==~ /^.*\/${readsLane2.getStemName()}_fastp.json/
            }
        }

    }

}
