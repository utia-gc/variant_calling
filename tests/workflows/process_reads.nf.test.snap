{
    "PROCESS_READS trims reads by default -- cutadapt + PE reads.": {
        "content": [
            [
                [
                    {
                        "sampleName": "SRR6924569",
                        "sampleNumber": "1",
                        "lane": "001",
                        "readType": "paired",
                        "trimStatus": "trimmed"
                    },
                    "SRR6924569_L001_trimmed_R1.fastq.gz:md5,1025b1a273fee2b51bd597f519ebbe86",
                    "SRR6924569_L001_trimmed_R2.fastq.gz:md5,3c3451e122dcf82b1be4a4526519fcc7"
                ],
                [
                    {
                        "sampleName": "SRR6924569",
                        "sampleNumber": "1",
                        "lane": "002",
                        "readType": "paired",
                        "trimStatus": "trimmed"
                    },
                    "SRR6924569_L002_trimmed_R1.fastq.gz:md5,0a8a869af34fd3f2c45b01d29b031c8f",
                    "SRR6924569_L002_trimmed_R2.fastq.gz:md5,5308d8b5e743b06434c3fa27776d34b6"
                ]
            ]
        ],
        "timestamp": "2023-10-19T13:29:30.171909727"
    },
    "PROCESS_READS trims reads by default -- cutadapt + SE reads.": {
        "content": [
            [
                [
                    {
                        "sampleName": "SRR1066657",
                        "sampleNumber": "3",
                        "lane": "001",
                        "readType": "single",
                        "trimStatus": "trimmed"
                    },
                    "SRR1066657_L001_trimmed_R1.fastq.gz:md5,0cf552582764f49d0eb969ffb0950ed4",
                    "SRR1066657_L001_trimmed_R2.NOFILE:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                [
                    {
                        "sampleName": "SRR1066657",
                        "sampleNumber": "3",
                        "lane": "002",
                        "readType": "single",
                        "trimStatus": "trimmed"
                    },
                    "SRR1066657_L002_trimmed_R1.fastq.gz:md5,c1ea5e2cbe17ba4c407d653af99fe2e9",
                    "SRR1066657_L002_trimmed_R2.NOFILE:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ]
        ],
        "timestamp": "2023-10-19T13:29:18.699389544"
    },
    "PROCESS_READS trims reads by default -- fastp +  PE reads.": {
        "content": [
            [
                [
                    {
                        "sampleName": "SRR6924569",
                        "sampleNumber": "1",
                        "lane": "001",
                        "readType": "paired",
                        "trimStatus": "trimmed"
                    },
                    "SRR6924569_L001_trimmed_R1.fastq.gz:md5,92338c4ea608209b461a6d747fc272bb",
                    "SRR6924569_L001_trimmed_R2.fastq.gz:md5,ffc0f8521e55d3f76548fc0da8b2f728"
                ],
                [
                    {
                        "sampleName": "SRR6924569",
                        "sampleNumber": "1",
                        "lane": "002",
                        "readType": "paired",
                        "trimStatus": "trimmed"
                    },
                    "SRR6924569_L002_trimmed_R1.fastq.gz:md5,9a0fa56bcd727f8b74e6be982dafe605",
                    "SRR6924569_L002_trimmed_R2.fastq.gz:md5,4f4ff327daa6b0c6e2daa2679c9ef894"
                ]
            ]
        ],
        "timestamp": "2023-10-31T13:44:10.791981151"
    },
    "PROCESS_READS trims reads by default -- fastp + SE reads.": {
        "content": [
            [
                [
                    {
                        "sampleName": "SRR1066657",
                        "sampleNumber": "3",
                        "lane": "001",
                        "readType": "single",
                        "trimStatus": "trimmed"
                    },
                    "SRR1066657_L001_trimmed_R1.fastq.gz:md5,fd66b28bfdc8dbc6888912de7bcebd92",
                    "SRR1066657_L001_trimmed_R2.NOFILE:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                [
                    {
                        "sampleName": "SRR1066657",
                        "sampleNumber": "3",
                        "lane": "002",
                        "readType": "single",
                        "trimStatus": "trimmed"
                    },
                    "SRR1066657_L002_trimmed_R1.fastq.gz:md5,450f10949d14764d502384fdc415f8e1",
                    "SRR1066657_L002_trimmed_R2.NOFILE:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ]
        ],
        "timestamp": "2023-10-31T13:43:53.539783212"
    }
}